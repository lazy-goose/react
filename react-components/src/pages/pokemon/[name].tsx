import { useRouter } from 'next/router';
import MainLayout from '@/components/MainLayout/MainLayout';
import PokeDetails from '@/components/MainLayout/PokeDetails/PokeDetails';
import PokeSearch from '@/components/MainLayout/PokeSearch/PokeSearch';

export default function RootComponent() {
  const router = useRouter();
  const pokemonName = (router.query.name as string) || '';

  return (
    <MainLayout
      main={<PokeSearch pokemonName={pokemonName} />}
      aside={<PokeDetails pokemonName={pokemonName} />}
    />
  );
}
