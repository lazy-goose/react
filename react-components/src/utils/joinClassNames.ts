export default function joinClassNames(
  ...classNames: (string | undefined | null)[]
) {
  return classNames.filter(Boolean).join(' ');
}
