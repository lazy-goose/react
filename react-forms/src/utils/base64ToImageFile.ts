async function base64ToImageFile(dataUrl: string, fileName: string): Promise<File> {
  const res: Response = await fetch(dataUrl);
  const blob: Blob = await res.blob();
  return new File([blob], fileName, { type: 'image/png' });
}

export default base64ToImageFile;
