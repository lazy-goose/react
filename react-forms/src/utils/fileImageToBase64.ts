async function fileImageToBase64(file: File) {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader();
    reader.addEventListener('error', () => reject(reader.error));
    reader.addEventListener('loadend', () => resolve(reader.result as string));
    reader.readAsDataURL(file);
  });
}

export default fileImageToBase64;
