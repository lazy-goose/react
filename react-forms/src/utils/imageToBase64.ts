async function imageToBase64(file: File) {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader();
    reader.addEventListener('error', () => reject(reader.error));
    reader.addEventListener('loadend', () => resolve(reader.result as string));
    reader.readAsDataURL(file);
  });
}

export default imageToBase64;
