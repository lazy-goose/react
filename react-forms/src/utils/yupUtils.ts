import { type AnySchema, type ValidationError } from 'yup';

type ErrorsObject = {
  [f: string]: string[];
};

type ExtractedErrorsObject = {
  [f: string]: string;
};

function yupErrorToObject(errors: ValidationError) {
  return errors.inner.reduce(
    (object, err) =>
      err.path ? { ...object, [err.path]: err.errors } : object,
    {} as ErrorsObject
  );
}

function extractFirstError(errors: ErrorsObject) {
  return Object.fromEntries(
    Object.entries(errors).map(([k, v]) => [k, v[0] || ''])
  ) as ExtractedErrorsObject;
}

async function validateSchema(
  schema: AnySchema,
  data: Record<string, unknown>
) {
  const filterNoLength = Object.fromEntries(
    Object.entries(data).filter(([, v]) => v !== '')
  );
  const yupErrors = await schema
    .validate(filterNoLength, { abortEarly: false })
    .catch((e) => e);
  return extractFirstError(yupErrorToObject(yupErrors));
}

export { yupErrorToObject, extractFirstError, validateSchema };
