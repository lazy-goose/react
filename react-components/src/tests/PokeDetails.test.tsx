import { describe, test, expect, vi } from 'vitest';
import { screen } from '@testing-library/react';
import renderComponent from './utils/renderComponent';
import userEvent from '@testing-library/user-event';
import pokemon from './data/pikachu.json';
import { MockPokemonRender } from './utils/mockComponents';
import mockRouter from 'next-router-mock';

describe('Tests for the Detailed Card component', () => {
  const renderPokemonPage = () =>
    renderComponent({
      children: <MockPokemonRender />,
      path: `/pokemon/${pokemon.name}`,
    });

  test('Check that a loading indicator is NOT displayed while fetching data', async () => {
    renderPokemonPage();
    const Loader = screen.queryByTestId('details-loader');
    expect(Loader).not.toBeInTheDocument();
  });

  test('Make sure the detailed card component correctly displays the detailed card data', async () => {
    renderPokemonPage();
    const Details = await screen.findByTestId('pokemon-details');
    await screen.findByTestId('pokemon-details-description');
    expect(Details).toMatchSnapshot();
  });

  test('Ensure that clicking the close button hides the component', async () => {
    const user = userEvent.setup();
    const { push } = mockRouter;
    mockRouter.push = vi.fn();
    renderPokemonPage();
    const CloseButton = await screen.findByTestId('close-aside');
    await user.click(CloseButton);
    expect(mockRouter.push).toBeCalled();
    mockRouter.push = push;
  });
});
