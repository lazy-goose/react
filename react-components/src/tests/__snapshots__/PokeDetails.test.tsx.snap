// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Detailed Card component > Make sure the detailed card component correctly displays the detailed card data 1`] = `
<div
  class="_PokeDetails_ad96d1"
  data-testid="pokemon-details"
>
  <h2
    class="_Heading_ad96d1"
  >
    pikachu
  </h2>
  <div
    class="_Container_ad96d1"
  >
    <ul
      data-testid="pokemon-details-description"
    >
      <li>
        <h4>
          Weight
        </h4>
        <p>
          60
        </p>
      </li>
      <li>
        <h4>
          height
        </h4>
        <p>
          4
        </p>
      </li>
      <li>
        <h4>
          Base Experience
        </h4>
        <p>
          112
        </p>
      </li>
      <li>
        <h4>
          Stats
        </h4>
        <ul>
          <li
            class="_KeyValue_ad96d1"
          >
            <span>
              hp
            </span>
            <span>
              35
            </span>
          </li>
          <li
            class="_KeyValue_ad96d1"
          >
            <span>
              attack
            </span>
            <span>
              55
            </span>
          </li>
          <li
            class="_KeyValue_ad96d1"
          >
            <span>
              defense
            </span>
            <span>
              40
            </span>
          </li>
          <li
            class="_KeyValue_ad96d1"
          >
            <span>
              special attack
            </span>
            <span>
              50
            </span>
          </li>
          <li
            class="_KeyValue_ad96d1"
          >
            <span>
              special defense
            </span>
            <span>
              50
            </span>
          </li>
          <li
            class="_KeyValue_ad96d1"
          >
            <span>
              speed
            </span>
            <span>
              90
            </span>
          </li>
        </ul>
      </li>
      <li>
        <h4>
          Abilities
        </h4>
        <ul>
          <li>
            static
          </li>
          <li>
            lightning-rod
          </li>
        </ul>
      </li>
      <li>
        <h4>
          Types
        </h4>
        <ul>
          <li>
            electric
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;
